cheatsheets/Cryptographic_Storage_Cheat_Sheet.md

```diff
@@ -1,234 +1,175 @@
+- A virtual HSM.
+- A secure storage such as [Amazon KMS](https://aws.amazon.com/kms/) or [Azure Key Vault](https://azure.microsoft.com/en-gb/services/key-vault/).
+- Secure storage APIs provided [ProtectedData](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.protecteddata?redirectedfrom=MSDN&view=netframework-4.8) class in the .NET framework.
 
-The requirement for split knowledge and/or dual control for key management prevents an individual user performing key management tasks such as key rotation or deletion. The system should require two individual users to perform an action (i.e. entering a value from their own OTP) which creates to separate values which are concatenated to create the final key data.
+In some cases such none of these will be available, such as in a shared hosting environment, meaning that it is not possible to obtain a high degree of protection for any encryption keys. However, the following basic rules can still be followed:
 
-###### 3.6.7 Prevention of unauthorized substitution of cryptographic keys
+- Do not hard-code keys into the application code.
+- Do not check keys into version control systems.
+- Protect the configuration files containing the keys with restrictive permissions.
+- Avoid storing keys in environment variables, as these can be accidentally exposed through functions such as [phpinfo()](https://www.php.net/manual/en/function.phpinfo.php).
 
-The system put in place to comply with requirement 3.6.6 can go a long way to preventing unauthorised substitution of key data. In addition to the dual control process you should implement strong access control, auditing and logging for key data so that unauthorised access attempts are prevented and logged.
+### Separation of Keys and Data
 
-###### 3.6.8 Requirement for cryptographic key custodians to sign a form stating that they understand and accept their key-custodian responsibilities
+Where possible, encryption keys should be stored in a separate location from encrypted data. For example, if the data is stored in a database, the keys should be stored in the filesystem. This means that if an attacker only has access to one of these (for example through directory traversal or SQL injection), they cannot access both the keys and the data.
 
-To perform the strong key management functions we have seen in requirement 3.6 we must have highly trusted and trained key custodians who understand how to perform key management duties. The key custodians must also sign a form stating they understand the responsibilities that come with this role.
+Depending on the architecture of the environment, it may be possible to store the keys and data on separate systems, which would provide a greater degree of isolation.
 
-# Related documentation & tools
+### Encrypting Stored Keys
 
-## Documentation
+Where possible, encryption keys should themselves be stored in an encrypted form. At least two separate keys are required for this:
```

Comment:
This is well described here: https://github.com/google/tink/blob/master/docs/JAVA-HOWTO.md#envelope-encryption.

We should mention why this is good approach (thread model)

Reply:
I've expanded this a bit and linked to some Google documentation talking about DEK/KEK usage patterns.